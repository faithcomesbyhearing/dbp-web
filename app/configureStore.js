/**
 * Create the store with dynamic reducers
 */

import { createStore, applyMiddleware, compose } from 'redux';
import { fromJS } from 'immutable';
import { routerMiddleware } from 'react-router-redux';
import createSagaMiddleware from 'redux-saga';
import {
	persistStore,
	autoRehydrate,
} from 'jh108-redux-persist-immutable-plugin';
import REDUX_PERSIST from './utils/reduxPersist';
import createReducer from './reducers';

const sagaMiddleware = createSagaMiddleware();

export default function configureStore(initialState = {}, history) {
	// Create the store with two middlewares
	// 1. sagaMiddleware: Makes redux-sagas work
	// 2. routerMiddleware: Syncs the location/URL path to the state
	const middlewares = [sagaMiddleware, routerMiddleware(history)];

	const enhancers = [applyMiddleware(...middlewares), autoRehydrate()];

	/* eslint-disable no-underscore-dangle */
	const composeEnhancers =
		process.env.NODE_ENV !== 'production' &&
		typeof window === 'object' &&
		window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
			? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({
					// Prevent recomputing reducers for `replaceReducer`
					shouldHotReload: false,
			  })
			: compose;
	/* eslint-enable */

	const store = createStore(
		createReducer(),
		fromJS(initialState),
		composeEnhancers(...enhancers),
	);

	if (typeof self === 'object') {
		persistStore(store, REDUX_PERSIST.storeConfig);
	}

	// Extensions
	store.runSaga = sagaMiddleware.run;
	store.injectedReducers = {}; // Reducer registry
	store.injectedSagas = {}; // Saga registry

	/* istanbul ignore next */
	if (module.hot) {
		module.hot.accept('./reducers', () => {
			store.replaceReducer(createReducer(store.injectedReducers));
		});
	}

	return store;
}
