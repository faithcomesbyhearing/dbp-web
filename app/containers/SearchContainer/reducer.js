/*
 *
 * SearchContainer reducer
 *
 */

import {
	ADD_SEARCH_TERM,
	GET_SEARCH_RESULTS,
	LOAD_SEARCH_RESULTS,
	SEARCH_ERROR,
	STOP_LOADING,
	VIEW_ERROR,
	START_LOADING,
} from './constants';

const initialState = structuredClone({
	searchResults: [],
	lastFiveSearches: [],
	trySearchOptions: [
		{ id: 1, searchText: 'Jesus' },
		{ id: 2, searchText: 'love' },
		{ id: 3, searchText: 'prayer' },
	],
	loadingResults: false,
	showError: false,
});

function searchContainerReducer(state = initialState, action = { type: null }) {
	switch (action.type) {
		case ADD_SEARCH_TERM: {
			const term = action.searchText.toLowerCase();

			if (state.lastFiveSearches.includes(term)) {
				return {
					...state,
					loadingResults: true,
				};
			}

			// build a brand-new array (never mutate state)
			const updated = [...state.lastFiveSearches, term];

			// keep only the last 9 entries
			if (updated.length > 9) {
				updated.shift();
			}

			return {
				...state,
				lastFiveSearches: updated,
				loadingResults: true,
			};
		}
		case GET_SEARCH_RESULTS:
			return {
				...state,
				loadingResults: true,
			};
		case LOAD_SEARCH_RESULTS:
			return {
				...state,
				loadingResults: false,
				showError: false,
				searchResults: structuredClone(action.searchResults),
			};
		case SEARCH_ERROR:
			return {
				...state,
				showError: true,
				loadingResults: false,
			};
		case VIEW_ERROR:
			return {
				...state,
				showError: false,
			};
		case STOP_LOADING:
			return {
				...state,
				loadingResults: false,
			};
		case START_LOADING:
			return {
				...state,
				loadingResults: true,
			};
		default:
			return state;
	}
}

export default searchContainerReducer;
