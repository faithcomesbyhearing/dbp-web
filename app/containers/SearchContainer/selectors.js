import { createSelector } from 'reselect';

/**
 * Direct selector to the searchContainer state domain
 */
const selectSearchContainerDomain = (state) => state.searchContainer;

/**
 * Other specific selectors
 */
const selectSearchResults = () =>
	createSelector(
		selectSearchContainerDomain,
		(search) => {
			const results = Array.isArray(search.searchResults)
				? search.searchResults
				: [];

			return results.reduce((acc, cur) => {
				const bookId = cur.book_id;
				// ensure we have an entry for this book
				if (!acc[bookId]) {
					acc[bookId] = {
						name: cur.book_name_alt,
						results: [],
					};
				}
				// push onto the _existing_ array
				acc[bookId].results.push(cur);
				return acc;
			}, {});
		},
	);

/**
 * Default selector used by SearchContainer
 */

const makeSelectSearchContainer = selectSearchContainerDomain;

export default makeSelectSearchContainer;
export { selectSearchContainerDomain, selectSearchResults };
