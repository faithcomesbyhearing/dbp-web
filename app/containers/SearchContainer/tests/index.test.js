import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

jest.mock('../../../utils/injectReducer', () => () => (Component) => Component);
jest.mock('../../../utils/injectSaga',   () => () => (Component) => Component);

import SearchContainer from '../index';
import {
	addSearchTerm,
	getSearchResults,
	stopLoading,
	startLoading,
	viewError,
} from '../actions';
import {
	ADD_SEARCH_TERM,
	GET_SEARCH_RESULTS,
	STOP_LOADING,
	START_LOADING,
	VIEW_ERROR,
} from '../constants';

const mockStore = configureStore([]);
const initialState = {
	searchContainer: {
		searchResults: [],
		lastFiveSearches: [],
		trySearchOptions: [
			{ id: 1, searchText: 'Jesus' },
			{ id: 2, searchText: 'love' },
			{ id: 3, searchText: 'prayer' },
		],
		loadingResults: false,
		showError: false,
	},
};
const store = mockStore(initialState);

describe('SearchContainer actions', () => {
	it('addSearchTerm should create an action with type ADD_SEARCH_TERM', () => {
		const payload = { bibleId: 'ENGESV', searchText: 'love' };
		const expected = { type: ADD_SEARCH_TERM, ...payload };
		expect(addSearchTerm(payload)).toEqual(expected);
	});

	it('getSearchResults should create an action with type GET_SEARCH_RESULTS', () => {
		const payload = { bibleId: 'ENGESV', searchText: 'peace' };
		const expected = { type: GET_SEARCH_RESULTS, ...payload };
		expect(getSearchResults(payload)).toEqual(expected);
	});

	it('stopLoading should create an action with type STOP_LOADING', () => {
		const payload = { foo: 'bar' };
		const expected = { type: STOP_LOADING, ...payload };
		expect(stopLoading(payload)).toEqual(expected);
	});

	it('startLoading should create an action with type START_LOADING', () => {
		const payload = { loading: true };
		const expected = { type: START_LOADING, ...payload };
		expect(startLoading(payload)).toEqual(expected);
	});

	it('viewError should create an action with type VIEW_ERROR', () => {
		const payload = { error: 'Something went wrong' };
		const expected = { type: VIEW_ERROR, ...payload };
		expect(viewError(payload)).toEqual(expected);
	});
	it('should render SearchContainer component properly', () => {
		const { asFragment } = render(
			<Provider store={store}>
				<SearchContainer
					searchText="fake search"
					bibleId="FAKE_BIBLE"
					onSearch={() => {}}
					loading={false}
					error={null}
				/>
			</Provider>,
		);
		expect(asFragment()).toMatchSnapshot();
	});
});
