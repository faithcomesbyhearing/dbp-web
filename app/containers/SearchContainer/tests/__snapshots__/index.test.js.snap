// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchContainer actions should render SearchContainer component properly 1`] = `
<DocumentFragment>
  <aside
    class="search-wrapper"
  >
    <aside
      class="search"
    >
      <header>
        <svg
          class="icon"
        >
          <use
            xlink:href="/svglist.svg#search"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          />
        </svg>
        <input
          class="input-class"
          placeholder="Search"
          value=""
        />
        <svg
          class="icon"
        >
          <use
            xlink:href="/svglist.svg#arrow_left"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          />
        </svg>
      </header>
      <div
        class="search-history-wrapper"
      >
        <div
          class="search-history"
        >
          <div
            class="starting"
          >
            <h2>
              Need a place to start? Try Searching:
            </h2>
            <button
              class="search-history-item"
              type="button"
            >
              Jesus
            </button>
            <button
              class="search-history-item"
              type="button"
            >
              love
            </button>
            <button
              class="search-history-item"
              type="button"
            >
              prayer
            </button>
          </div>
          <div
            class="history"
          >
            <div
              class="header"
            >
              <h2>
                Search History:
              </h2>
            </div>
          </div>
        </div>
      </div>
    </aside>
  </aside>
</DocumentFragment>
`;
